socs/agents/hwp_supervisor/agent.py

```diff
@@ -750,6 +795,89 @@ def sleep_until_complete(self, session=None, dt=1):
             time.sleep(dt)
 
 
+@contextmanager
+def ensure_grip_safety(hwp_state: HWPState, timeout: float = 60.):
+    """
+    Run required checks for gripper safety. This will check ACU parameters such
+    as az/el position and velocity. If `use_acu_blocking` is set, this will
+    also attempt to block out ACU motion for the duration of the operation.
+
+    Args
+    ------
+    hwp_state : HWPState
+        HWP state object. In addition to reading ACU state vars, this will
+        set the `request_block_ACU_motion` flag if `use_acu_blocking` is set.
+    timeout: float
+        Timeout for waiting for the ACU blockout before an error will be raised.
+    """
+    if hwp_state.pid_current_freq is None:
+        raise RuntimeError("Cannot determine current HWP Freq")
+
+    if np.abs(hwp_state.pid_current_freq) > 0.02:
+        raise RuntimeError("Cannot grip HWP while spinning")
+
+    acu = hwp_state.acu
+    if acu is None:
+        yield
+        return
```

Comment:
I'm concerned about this logic. At least we should have a warning?

Reply:
I'm also wondering about the behavior here. Does this just skip the position checks for the ACU? Doesn't that mean we can end up in an unsafe position?

Reply:
Yes, so I don't think an ACU should be required to operate the HWP supervisor, and when an ACU is not present I think skipping checks (without printing warning logs every check) is the correct behavior. Currently, I do log an info statement [here](https://github.com/simonsobs/socs/blob/811858ad075671b2306c16e9a3edcfff579c8797/socs/agents/hwp_supervisor/agent.py#L310) but this could definitely be hard to catch if your not looking for it...

What if we add an additional command line argument `--no-acu` which needs to be set to bypass ACU checks? Then I can throw an error if `--acu-instance-id` is not passed without the `--no-acu` flag, to prevent SATs from accidentally forgetting to set the instance id.

Reply:
Yeah, reading [this Instrument Safety page](https://simonsobs.atlassian.net/wiki/spaces/PRO/pages/433684484/CHWP+commissioning#Instrument-Safety) linked from https://github.com/simonsobs/daq-discussions/discussions/96, it does seem like the only relevant thing for grip/ungrip is not to move the platform while doing it. So I agree that not defining the ACU shouldn't stop this from working.

> What if we add an additional command line argument `--no-acu` which needs to be set to bypass ACU checks? Then I can throw an error if `--acu-instance-id` is not passed without the `--no-acu` flag, to prevent SATs from accidentally forgetting to set the instance id.

This sounds okay. If platforms are going to forget to add `--acu-instance-id`, they'll probably also forget to add `--no-acu` -- so I guess default to requiring checks. I wouldn't be against setting the default `--acu-instance-id` to 'acu' either, as that's the instance-id across all platforms, and I don't really see that changing -- if we really are going to forget to set that flag after this is merged.

Reply:
Let me ask about this.
Do you think there is possibilities that hwp_supervisor tries to grip hwp while telescope is moving or fails to block acu motion? If acu is None, the telescope is not moving and will not move later?
Also, can hwp_supervisor perform safety checks when acu crashes and it is restarted later?

Reply:
`acu` will only ever be None if `the `--no-acu` flag is set in the site-config file. If an ACU instance id is set, but the agent is not found or not responding, I believe one of the other checks will fail and you will be unable to grip with the supervisor until the ACU agent is working again.

At the moment I'm not sure if we want to allow the supervisor to grip when the ACU agent is not responding. I think we should not allow it for the time being (since there are other methods to grip) and if it is a problem we can always come back and change the behavior.

Reply:
I see. Thanks for the explanation.